### DISCUSSION TITLE
New operator ?:: to be introduced in C#

### DISCUSSION THREAD

**Discussion by kasimhusaini** [dotnet/csharplang#3632] [Language Ideas]
The operator ?: takes a boolean value on its left side. The operator ?:: can take boolean? [Nullable boolean value] on its left side.

It should be something like this:
(bool?)?<true value>:<false value>:<null value>

**Comment by 333fred** [comment]
What advantages would this have over a simple switch expression?

**Comment by juliusfriedman** [comment]
See also:

https://github.com/juliusfriedman/net7mma_core/blob/master/Common/Classes/Binary/BitReader.cs#L312

```c#
  public bool PeekBit(bool reverse = false)
        {
            switch (reverse) 
            {
                default://Since .net core 3.1 c# compiler requires this or the build error is that this method doesn't return on all code paths....
                case false: return Common.Binary.GetBitReverse(ref m_ByteCache.Array[m_ByteIndex], m_BitIndex);
                case true: return Common.Binary.GetBit(ref m_ByteCache.Array[m_ByteIndex], m_BitIndex);
            }
        }
```
I can no longer find the issue I opened for this where I was told by @gafter that this was by design due a bug report where @jkotas initially cc'd him.

**Comment by theunrepentantgeek** [comment]
What order are the three elements?

For `?:` we know they are **true** then **false**, from analogy with the **if** statement.

For `?::` there are three possibilities that are consistent with this: 

* **true**/**false**/**null**
* **null**/**true**/**false**
* ~**true**/**null**/**false**~

Ok, make it two ... that last one is just silly.

But seriously, what order would they be? And why? And how would millions of C# developers know?

Keep in mind that many experienced C# developers never read the docs for new versions, they pick up new syntax by seeing it in use. (In my experiences, this applies to almost all C# developers; I suspect those who actively seek to learn new language features are a terribly small minority.)

Also, your proposed syntax is ambiguous:

``` csharp
return eatsMeat ? eatsPlants : sheep : monkey : tiger;
```
Is this a `?::` inside a `?:` or is it a `?:` inside a `?::`?

**Comment by theunrepentantgeek** [comment]
@juliusfriedman, at the risk of being off topic (since the OP is about `bool?`, which you don't use in your example), there are at least two ways to express that more simply:

Using the new _switch expression:_
```
public bool PeekBit(bool reverse = false)
    => reverse switch {
        false => Common.Binary.GetBitReverse(ref m_ByteCache.Array[m_ByteIndex], m_BitIndex),
        true => Common.Binary.GetBit(ref m_ByteCache.Array[m_ByteIndex], m_BitIndex),
    };
```

or the classic approach:
```
public bool PeekBit(bool reverse = false)
{
    if (reverse) 
    {
        return Common.Binary.GetBit(ref m_ByteCache.Array[m_ByteIndex], m_BitIndex);
    }

    return Common.Binary.GetBitReverse(ref m_ByteCache.Array[m_ByteIndex], m_BitIndex);
}
```

**Comment by miloush** [comment]
@theunrepentantgeek pretty convinced most developers would say the order is true/false/null. Taking intuitiveness aside, it is superset of the existing syntax (you just add stuff to the right), and it is also the defined descending order of bool? values.

I do see a value in the shorter syntax compared to a switch expression... 
```C#
public Brush IsCheckedBrush => _checkBox.IsChecked ? Brushes.Green : Brushes.Red : Brushes.Yellow;
```

But I am not sure the expected usage is worth cost.

**Comment by juliusfriedman** [comment]
@miloush, I find that syntax hard to `read` Yes I could get my head around it but we already know:

```c#
public Brush IsCheckedBrush => _checkBox.IsChecked ? Brushes.Green : _checkBox.IsChecked ? Brushes.Red : Brushes.Yellow;
```

So what is this really saving?  This really would only be optimal if the multiple checks of `_checkBox.IsChecked` were removed, and for the person who didn't want that behavior (due to threading or otherwise) would be stuck in the write it yourself boat (use a `switch/lea` or multiple `cmp/cmov`) e.g. as I did above.

**Comment by juliusfriedman** [comment]
@theunrepentantgeek
> @juliusfriedman, at the risk of being off topic (since the OP is about `bool?`, which you don't use in your example), there are at least two ways to express that more simply:
> 
> Using the new _switch expression:_
> 
> ```
> public bool PeekBit(bool reverse = false)
>     => reverse switch {
>         false => Common.Binary.GetBitReverse(ref m_ByteCache.Array[m_ByteIndex], m_BitIndex),
>         true => Common.Binary.GetBit(ref m_ByteCache.Array[m_ByteIndex], m_BitIndex),
>     };
> ```
> 
> or the classic approach:
> 
> ```
> public bool PeekBit(bool reverse = false)
> {
>     if (reverse) 
>     {
>         return Common.Binary.GetBit(ref m_ByteCache.Array[m_ByteIndex], m_BitIndex);
>     }
> 
>     return Common.Binary.GetBitReverse(ref m_ByteCache.Array[m_ByteIndex], m_BitIndex);
> }
> ```

What I don't agree with is that the new switch statement allows a non exhaustive list and is therefore not complete.

The classic approach is fine as there is only 2 cases I can observe there anyway and could also be written as:

```c#
 public bool PeekBit(bool reverse = false)
{
    return reverse ?
    Common.Binary.GetBit(ref m_ByteCache.Array[m_ByteIndex], m_BitIndex);
    :
    Common.Binary.GetBitReverse(ref m_ByteCache.Array[m_ByteIndex], m_BitIndex);
}
```

This is why I disagree that the normal `switch` should be treated any differently.

And while I know this is about `bool?` this why the normal `switch` was changed so thus I feel like it's not that far off topic.

**Comment by miloush** [comment]
@juliusfriedman I am not sure I understand your example - it does not compile, since `bool?` cannot be used in this operator. You could do
```C#
public Brush IsCheckedBrush => _checkBox.IsChecked  == true ? Brushes.Green : _checkBox.IsChecked == false ? Brushes.Red : Brushes.Yellow;
```

Yes I am not claiming it is the easiest thing to read, but it is better than above and as I said I don't expect this to be read very often. If you are not bothered with extra evaluation, then you can just write your own method.

**Comment by juliusfriedman** [comment]
@miloush, sorry I think i missing a few of the `_checkBox?.IsChecked ??` `null coalescing operators` ;) otherwise what does it return not `bool?`? What am I missing?

**Comment by miloush** [comment]
@juliusfriedman `??` needs to return `bool` in this case, so you cannot use it to return different type. Anyway, here is a trick with single evaluation and almost the same syntax:
```C#
public Brush IsCheckedBrush => _checkBox.IsChecked is bool b ? b ? Brushes.Green : Brushes.Red : Brushes.Yellow;
```

Yeah, now I am even less convinced this is needed. :)

**Comment by jnm2** [comment]
Or, what I think is clearer:

```cs
public Brush IsCheckedBrush =>
    _checkBox.IsChecked switch { true => Brushes.Green, false => Brushes.Red, _ => Brushes.Yellow };
```

**Comment by juliusfriedman** [comment]
Just for the record:

https://sharplab.io/#gist:bc59d7130bdc46cce9e0f55a51faaa3e

The example @miloush posted is 8 bytes shorter.

**Comment by jnm2** [comment]
@juliusfriedman That's only in debug compilations. [It goes the other way](https://sharplab.io/#v2:C4LglgNgPgAgTARgFAwMwAJ7oMLoN5LpHoBGA9mRAPzrACmAzsOgLzoB2ArhBANyHE06MO2YBJBtgAWdAMYBrOgBMAQgCdODKawB8tRszANSFCKXQ0SF9AnQg4d9Kn7FMGEeMkyFy9ZqkAygDuYMCy2iw6AkT0TOgMIWHaeLQadLo2ADToAGYAhhAM6ZFZ6AD6GRgAvvxVQA) in release.

**Comment by juliusfriedman** [comment]
@jnm2 Thank you for the insight!
